File: src/main/java/me/redepicness/bingo/world/EmptyChunkGenerator.java
package me.redepicness.bingo.world;

import org.bukkit.HeightMap;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.generator.BiomeProvider;
import org.bukkit.generator.BlockPopulator;
import org.bukkit.generator.ChunkGenerator;
import org.bukkit.generator.WorldInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Random;

import static me.redepicness.bingo.world.VoidBiomeProvider.VOID_BIOME_PROVIDER;

public class EmptyChunkGenerator extends ChunkGenerator {

    private final int platformRadius;
    private final int chunkRadius;

    public EmptyChunkGenerator(int platformRadius) {
        this.platformRadius = platformRadius;
        this.chunkRadius = (platformRadius / 16) + 1;
    }

    @Override
    public void generateNoise(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkData chunkData) {
        if (Math.abs(chunkX) > chunkRadius || Math.abs(chunkZ) > chunkRadius) {
            return;
        }

        int safeY = getSafeY(worldInfo.getMinHeight(), worldInfo.getMaxHeight());

        for (int x = 0; x < 16; x++) {
            for (int z = 0; z < 16; z++) {
                int absX = Math.abs(chunkX * 16 + x);
                int absY = Math.abs(chunkZ * 16 + z);

                if (absX <= platformRadius && absY <= platformRadius) {
                    chunkData.setBlock(x, safeY, z, Material.STONE);
                }
            }
        }
        if (chunkX == 0 && chunkZ == 0) {
            chunkData.setBlock(0, safeY, 0, Material.BEDROCK);
        }
    }

    @Override
    public void generateSurface(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkData chunkData) {}

    @Override
    public void generateBedrock(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkData chunkData) {}

    @Override
    public void generateCaves(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ, @NotNull ChunkData chunkData) {}

    @Override
    public @Nullable BiomeProvider getDefaultBiomeProvider(@NotNull WorldInfo worldInfo) {
        return VOID_BIOME_PROVIDER;
    }

    @Override
    public int getBaseHeight(@NotNull WorldInfo worldInfo, @NotNull Random random, int x, int z, @NotNull HeightMap heightMap) {
        if (Math.abs(x) <= platformRadius && Math.abs(z) <= platformRadius) {
            return getSafeY(worldInfo.getMinHeight(), worldInfo.getMaxHeight());
        } else {
            return worldInfo.getMinHeight();
        }
    }

    @Override
    public boolean canSpawn(@NotNull World world, int x, int z) {
        return Math.abs(x) <= platformRadius && Math.abs(z) <= platformRadius;
    }

    @Override
    public @NotNull List<BlockPopulator> getDefaultPopulators(@NotNull World world) {
        return Collections.emptyList();
    }

    @Override
    public @Nullable Location getFixedSpawnLocation(@NotNull World world, @NotNull Random random) {
        return new Location(world, 0, getSafeY(world.getMinHeight(), world.getMaxHeight()) + 1, 0);
    }

    @Override
    public boolean shouldGenerateNoise() {
        return false;
    }

    @Override
    public boolean shouldGenerateNoise(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ) {
        return false;
    }

    @Override
    public boolean shouldGenerateSurface() {
        return false;
    }

    @Override
    public boolean shouldGenerateSurface(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ) {
        return false;
    }

    @Override
    public boolean shouldGenerateCaves() {
        return false;
    }

    @Override
    public boolean shouldGenerateCaves(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ) {
        return false;
    }

    @Override
    public boolean shouldGenerateDecorations() {
        return false;
    }

    @Override
    public boolean shouldGenerateDecorations(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ) {
        return false;
    }

    @Override
    public boolean shouldGenerateMobs() {
        return false;
    }

    @Override
    public boolean shouldGenerateMobs(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ) {
        return false;
    }

    @Override
    public boolean shouldGenerateStructures() {
        return false;
    }

    @Override
    public boolean shouldGenerateStructures(@NotNull WorldInfo worldInfo, @NotNull Random random, int chunkX, int chunkZ) {
        return false;
    }

    private int getSafeY(int minHeight, int maxHeight) {
        return Math.max(minHeight, Math.min(99, maxHeight - 1));
    }

}

File: src/main/kotlin/me/redepicness/bingo/world/generation/EmptyChunkGenerator.kt
package me.redepicness.bingo.world.generation

import org.bukkit.HeightMap
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.generator.BiomeProvider
import org.bukkit.generator.BlockPopulator
import org.bukkit.generator.ChunkGenerator
import org.bukkit.generator.WorldInfo
import java.util.*
import kotlin.math.abs
import kotlin.math.absoluteValue

class EmptyChunkGenerator(private val platformRadius: Int) : ChunkGenerator() {

    private val chunkRadius: Int = platformRadius / 16 + 1

    override fun generateNoise(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int, chunkData: ChunkData) {
        if (chunkX.absoluteValue > chunkRadius || chunkZ.absoluteValue > chunkRadius) {
            return
        }
        val safeY = getSafeY(worldInfo.minHeight, worldInfo.maxHeight)
        for (x in 0..15) {
            for (z in 0..15) {
                val absX = abs(chunkX * 16 + x)
                val absY = abs(chunkZ * 16 + z)
                if (absX <= platformRadius && absY <= platformRadius) {
                    chunkData.setBlock(x, safeY, z, Material.STONE)
                }
            }
        }
        if (chunkX == 0 && chunkZ == 0) {
            chunkData.setBlock(0, safeY, 0, Material.BEDROCK)
        }
    }

    override fun generateSurface(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int, chunkData: ChunkData) {}
    override fun generateBedrock(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int, chunkData: ChunkData) {}
    override fun generateCaves(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int, chunkData: ChunkData) {}
    override fun getDefaultBiomeProvider(worldInfo: WorldInfo): BiomeProvider {
        return VoidBiomeProvider
    }

    override fun getBaseHeight(worldInfo: WorldInfo, random: Random, x: Int, z: Int, heightMap: HeightMap): Int {
        return if (x.absoluteValue <= platformRadius && z.absoluteValue <= platformRadius) {
            getSafeY(worldInfo.minHeight, worldInfo.maxHeight)
        } else {
            worldInfo.minHeight
        }
    }

    override fun canSpawn(world: World, x: Int, z: Int): Boolean {
        return x.absoluteValue <= platformRadius && z.absoluteValue <= platformRadius
    }

    override fun getDefaultPopulators(world: World): List<BlockPopulator> {
        return emptyList()
    }

    override fun getFixedSpawnLocation(world: World, random: Random): Location? {
        return Location(world, 0.0, (getSafeY(world.minHeight, world.maxHeight) + 1).toDouble(), 0.0)
    }

    override fun shouldGenerateNoise(): Boolean {
        return false
    }

    override fun shouldGenerateNoise(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int): Boolean {
        return false
    }

    override fun shouldGenerateSurface(): Boolean {
        return false
    }

    override fun shouldGenerateSurface(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int): Boolean {
        return false
    }

    override fun shouldGenerateCaves(): Boolean {
        return false
    }

    override fun shouldGenerateCaves(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int): Boolean {
        return false
    }

    override fun shouldGenerateDecorations(): Boolean {
        return false
    }

    override fun shouldGenerateDecorations(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int): Boolean {
        return false
    }

    override fun shouldGenerateMobs(): Boolean {
        return false
    }

    override fun shouldGenerateMobs(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int): Boolean {
        return false
    }

    override fun shouldGenerateStructures(): Boolean {
        return false
    }

    override fun shouldGenerateStructures(worldInfo: WorldInfo, random: Random, chunkX: Int, chunkZ: Int): Boolean {
        return false
    }

    private fun getSafeY(minHeight: Int, maxHeight: Int): Int {
        return 99.coerceIn(minHeight..<maxHeight) // Math.max(minHeight, 99.coerceAtMost(maxHeight -1))
    }
}
